### ISSUE TITLE
Ember Inspector causing "Assertion Failed: Illegal set of identifier"

### ISSUE THREAD

**Issue by simonc** [emberjs/ember-inspector#1875]
**Describe the bug**

After experiencing random cases of the `Assertion Failed: Illegal set of identifier` error I looked for any information and ended up on [this StackOverflow question](https://stackoverflow.com/questions/69262963/emberjs-data-assert-in-setrecorddatafor-illegal-set-of-identifier) and more specifically [this comment by HappyDude](https://stackoverflow.com/questions/69262963/emberjs-data-assert-in-setrecorddatafor-illegal-set-of-identifier#comment124776510_69262963) stating that this error can sometimes be caused by Ember Inspector.

I disabled Ember Inspector and it fixed the issue.

One element I can add is that the model causing the issue appeared twice in the inspector Data tab.

**To Reproduce**

I'm very sad and sorry not to be able to give any useful reproduction steps as it happens randomly.

**Environment**

Ember: 3.28.8
Ember Data: 3.28.6
Ember Inspector: 4.5.11
Firefox: 95.0.2 (64-bit) on a Mac

---

Taking the opportunity to thank all the team working on this otherwise wonderful tool that Ember Inspector is ❤️

**Comment by numero-trey** [comment]
I just ran into this bug, and it happened when I added a hasMany to the model I was working with.

**Comment by rwwagner90** [comment]
@numero-trey would you be able to provide us with a small reproduction please?

**Comment by lifeart** [comment]
same for me, preconditions:

ember-data: 3.28.7
browser: chrome
devtools open: true
ember inspector open: true
ember-data request: findRecord
called in model hook

**Comment by luxferresum** [comment]
Same for me. I was able to [create a reproduction](https://github.com/luxferresum/illegal-set-of-identifier-reproduction).

**Comment by lifeart** [comment]
closes callback (before error)

![image](https://user-images.githubusercontent.com/1360552/151066837-8e763ec1-2ca9-40d8-b608-bb37ae2d358e.png)

Looks like error happen inside "fetch {modelType}" buckburner queue

**Comment by lifeart** [comment]
error firing in this line:
![image](https://user-images.githubusercontent.com/1360552/151070809-a2506f41-54f9-41f9-a6b0-31842770c487.png)


also, after error, model is unloaded  (in catch case) =>

![image](https://user-images.githubusercontent.com/1360552/151070891-ea00a48c-7e59-4409-9b08-82fcd474bde8.png)

**Comment by lifeart** [comment]
tldr: looks like `observeModelType` called 2 times, and it's leads to error

**Comment by lifeart** [comment]
looks like we have idea how to fix it: https://discord.com/channels/480462759797063690/486549196837486592/935672357840900118

**Comment by lifeart** [comment]
![image](https://user-images.githubusercontent.com/1360552/151317069-b81135ef-644d-49ef-bee9-cd46a4151c17.png)
if I comment `iterate(records, () => {})` error does not appear

**Comment by rwwagner90** [comment]
@lifeart looks like from the comments there this might just be a leftover legacy thing?

**Comment by lifeart** [comment]
Looks like it's still used. :/

**Comment by csprocket777** [comment]
I'm experiencing this today with the same environment. Any guidance on this?

**Comment by lifeart** [comment]
@csprocket777 try to load all model types before opening ember-inspector

**Comment by Techn1x** [comment]
burned a few hours trying to figure out the cause of this one as well :(

it's a bit of a double whammy because the tools you use to debug, like ember inspector, are the last thing you expect to be causing issues

**Comment by patricklx** [comment]
Does https://github.com/emberjs/data/pull fix this issue too for anyone? I did not see it anymore.

**Comment by mcfiredrill** [comment]
> Does https://github.com/emberjs/data/pull/8008 fix this issue too for anyone? I did not see it anymore.

I've been bitten by this multiple times, if this fix works I'd appreciate a backport to 3.2.x series. :pray:

**Comment by patricklx** [comment]
fhis has been fixed now in ember and ember data. The ember data fix has also been backported. The fix in ember might also be required. If you still see this issue comment on https://github.com/emberjs/ember.js/pull/20127 for a backport
